functions/src/validation.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  body,
+  query,
+  ValidationError,
+  validationResult,
+} from "express-validator";
+import { NextFunction, Request, Response } from "express";
+import { APIFields, isValidUrl, ShareLinksCollection } from "./utils";
+import {
+  getShareLinkErrorByCode,
+  sendAndThrowShareLinkErrorCode,
+  ShareLinkErrorCode,
+} from "./error-handling";
+import * as functions from "firebase-functions";
+
+export function registerUrlValidationRules() {
+  return [
+    // Use a custom URL validator b/c we need to ensure an HTTP(S) protocol is
+    // present for use in the /go/ endpoint.
+    body(ShareLinksCollection.URL).custom(isValidUrl),
+    body(ShareLinksCollection.IMAGE_URL).optional().custom(isValidUrl),
+    body(ShareLinksCollection.TITLE).optional().isString(),
+    body(ShareLinksCollection.DESCRIPTION).optional().isString(),
+    body(ShareLinksCollection.IMAGE_HEIGHT).optional().isInt(),
+    body(ShareLinksCollection.IMAGE_WIDTH).optional().isInt(),
+  ];
+}
+
+export function createApiKeyValidationRules() {
+  return [body(APIFields.EMAIL).isEmail()];
+}
+
+export function modifyApiKeyValidationRules() {
+  return [body(APIFields.EMAIL).isEmail(), body(APIFields.ENABLED).isBoolean()];
+}
+
+export function queryUrlValidationRule() {
+  return [query(ShareLinksCollection.URL).custom(isValidUrl)];
+}
+
+export function validate(req: Request, res: Response, next: NextFunction) {
+  const errors = validationResult(req);
+  if (errors.isEmpty()) {
+    return next();
+  }
+  const error = errors.array()[0]; // Just handle the first error if there are multiple.
+  return validationToShareLinkError(res, error);
+}
+
+function validationToShareLinkError(
```

Comment:
Was on the fence here about just handling the validation errors without using the `ShareLinkErrorCode` catalog (e.g. just sending something like `.status(422).send(${error.msg}. ${error.param}: ${error.value})`), but it felt better to try to keep the errors as centralized as possible, even though it adds a little bit of plumbing/handling

Reply:
This function sounds like it would just return a share link error, but it will send and (usually) throw the error as well.  Can we rename it sendAndThrowValidationError() or something?